<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Simulating epidemics with the toyepidemic R-package • toyepidemic</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">toyepidemic</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/UserGuide.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Simulating epidemics with the toyepidemic R-package</h1>
                        <h4 class="author">Venelin Mitov</h4>
            
          </div>

    
    
<div class="contents">
<div id="designing-the-epidemic" class="section level1">
<h1 class="hasAnchor">
<a href="#designing-the-epidemic" class="anchor"></a>Designing the epidemic</h1>
<div id="define-the-pathogen-genotypes" class="section level2">
<h2 class="hasAnchor">
<a href="#define-the-pathogen-genotypes" class="anchor"></a>Define the pathogen genotypes</h2>
<p>We set the numbers of different alleles for each quantitative trait locus (QTL) in the pathogen genotype. These are specified in the form of a integer vector with elements bigger or equal to 2. The length of this vector corresponds to the number of QTLs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># numbers of alleles for each quantitative trait locus</span>
numsAllelesAtSites &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">3</span>, <span class="dv">2</span>)

<span class="co"># Define a matrix of the possible genotype encodings (as allele contents).</span>
genotypes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateGenotypes.html">generateGenotypes</a></span>(numsAllelesAtSites)
<span class="kw">print</span>(genotypes)</code></pre></div>
<pre><code>##      [,1] [,2]
## [1,]    1    1
## [2,]    1    2
## [3,]    2    1
## [4,]    2    2
## [5,]    3    1
## [6,]    3    2</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Define the probability of the first infecting strain (the first pathogen strain starting the epidemic)</span>
pg.init &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">0</span>, <span class="kw">nrow</span>(genotypes))
pg.init[<span class="dv">1</span>] &lt;-<span class="st"> </span><span class="dv">1</span> <span class="co"># we specify that 1 is the first strain with probability 1.</span></code></pre></div>
</div>
<div id="define-the-number-and-the-frequencies-of-host-types" class="section level2">
<h2 class="hasAnchor">
<a href="#define-the-number-and-the-frequencies-of-host-types" class="anchor"></a>Define the number and the frequencies of host-types</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># number of host-types</span>
n &lt;-<span class="st"> </span><span class="dv">6</span>

<span class="co"># probability of each host type in a susceptible compartment </span>
pe &lt;-<span class="st"> </span><span class="kw">runif</span>(n)
pe &lt;-<span class="st"> </span>pe<span class="op">/</span><span class="kw">sum</span>(pe) <span class="co"># ensure that they sum-up to 1</span></code></pre></div>
</div>
<div id="define-the-ge-values" class="section level2">
<h2 class="hasAnchor">
<a href="#define-the-ge-values" class="anchor"></a>Define the GE-values</h2>
<p>The GE-values represent the general genotype x host-type effects. A genotype x host-type effect is the mean trait value of an individual of a given host-type carrying a given pathogen strain in the absence of selection.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># General host-type x strain effects (expected phenotypes for genotype by environment combinations)</span>
GEVs &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="ot">NA</span>, <span class="dt">nrow=</span>n, <span class="dt">ncol=</span><span class="kw">nrow</span>(genotypes))

<span class="co"># assign random values to each genotype x host-type combination:</span>
<span class="cf">for</span>(g <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(genotypes)) {
  GEVs[, g] &lt;-<span class="st"> </span><span class="kw">rnorm</span>(<span class="dt">n=</span>n, <span class="dt">mean=</span><span class="dv">2</span><span class="op">+</span><span class="dv">2</span><span class="op">/</span><span class="kw">nrow</span>(genotypes)<span class="op">*</span>(g), <span class="dt">sd=</span><span class="fl">0.4</span>)
}</code></pre></div>
<p>To get a visual idea of the different GE-values, we can plot them in the form of an R-matplot:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">matplot</span>(GEVs, <span class="dt">type=</span><span class="st">'l'</span>, <span class="dt">lty=</span><span class="dv">1</span>, <span class="dt">col=</span><span class="dv">1</span><span class="op">:</span><span class="dv">6</span>)</code></pre></div>
<p><img src="UserGuide_files/figure-html/unnamed-chunk-4-1.png" width="672"></p>
</div>
<div id="define-the-host-specific-effects-distribution" class="section level2">
<h2 class="hasAnchor">
<a href="#define-the-host-specific-effects-distribution" class="anchor"></a>Define the host-specific effects distribution</h2>
<p>The final phenotype of every infected host is the sum of the GE-value for its currently infecting strain and a random host-specific effect. The toyepidemic package assumes that the host-specific effect is a normally distributed random variable drawn exactly once for each possible infecting strain in a newly infected host. This normal distribution has mean 0 and standard deviation defined by the following parameter <span class="math inline">\(\sigma_e\)</span>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># is the special environmental effect unique for each pathogen genoetype in an individual</span>
eUniqForEachG &lt;-<span class="st"> </span><span class="ot">TRUE</span>

sigmae &lt;-<span class="st"> </span>.<span class="dv">6</span>
<span class="co"># it is possible to specify different standard deviations for the different host-types, i.e. different host-types exhibit stronger or weaker effect on the trait. For this example we keep them fixed for all host-types.</span>
sde &lt;-<span class="st"> </span><span class="kw">rep</span>(sigmae, n)</code></pre></div>
</div>
<div id="defining-the-population-parameters" class="section level2">
<h2 class="hasAnchor">
<a href="#defining-the-population-parameters" class="anchor"></a>Defining the population parameters</h2>
<p>At the between-host level we start by defining the size, the birth- and the natural death rate in the population at equilibrium. These are constants:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#initial population size (equilibrium in the absence of disease)</span>
N &lt;-<span class="st"> </span><span class="fl">1e5</span>

<span class="co"># setting the between-host and within-host dynamics of the simulation:</span>
<span class="co"># natural death rate</span>
mu &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">/</span><span class="dv">850</span>

<span class="co"># constant birth rate that maintains this equilibrium</span>
nu &lt;-<span class="st"> </span><span class="kw">ifelse</span>(<span class="kw">is.finite</span>(N), mu<span class="op">*</span>N, <span class="dv">0</span>)</code></pre></div>
<p>Next, we define the rate parameters for infected hosts: ### Infected death-rate</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># death-rate as a function of viral load and natural death rate mu</span>
rateDie &lt;-<span class="st"> </span><span class="cf">function</span>(z, mu) {
  V &lt;-<span class="st"> </span><span class="dv">10</span><span class="op">^</span>z
  Dmin &lt;-<span class="st"> </span><span class="dv">2</span>
  Dmax &lt;-<span class="st"> </span><span class="dv">25</span><span class="op">*</span><span class="dv">12</span>
  D50 &lt;-<span class="st"> </span><span class="dv">10</span><span class="op">^</span><span class="dv">3</span>
  Dk &lt;-<span class="st"> </span><span class="fl">1.4</span>
  (V<span class="op">^</span>Dk<span class="op">+</span>D50<span class="op">^</span>Dk)<span class="op">/</span>(Dmin<span class="op">*</span>(V<span class="op">^</span>Dk<span class="op">+</span>D50<span class="op">^</span>Dk)<span class="op">+</span>((Dmax<span class="op">-</span>Dmin)<span class="op">*</span>D50<span class="op">^</span>Dk)) <span class="op">+</span><span class="st"> </span>mu
}

<span class="kw">curve</span>(<span class="kw">rateDie</span>(x, mu), <span class="dv">0</span>, <span class="dv">7</span>, <span class="dt">col=</span><span class="st">"red"</span>)</code></pre></div>
<p><img src="UserGuide_files/figure-html/unnamed-chunk-7-1.png" width="336"></p>
<div id="transmission-rate-as-a-function-of-viral-load-and-rate-of-risky-contacts-with-susceptible-hosts" class="section level3">
<h3 class="hasAnchor">
<a href="#transmission-rate-as-a-function-of-viral-load-and-rate-of-risky-contacts-with-susceptible-hosts" class="anchor"></a>Transmission-rate as a function of viral load and rate of risky contacts with susceptible hosts</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rateContact &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">/</span><span class="dv">6</span>

<span class="co"># note that at runtime, rateInfect is scaled by the proportion of susceptible hosts</span>
rateInfect &lt;-<span class="st"> </span><span class="cf">function</span>(z, rateContact) {
  V &lt;-<span class="st"> </span><span class="dv">10</span><span class="op">^</span>z
  Emin &lt;-<span class="st"> </span>.<span class="dv">3</span>
  Emax &lt;-<span class="st"> </span>.<span class="dv">6</span>
  E50 &lt;-<span class="st"> </span><span class="dv">10</span><span class="op">^</span><span class="dv">3</span>
  Ek &lt;-<span class="st"> </span><span class="fl">1.4</span>
  E &lt;-<span class="st"> </span>Emin<span class="op">+</span>(Emax<span class="op">-</span>Emin)<span class="op">*</span>V<span class="op">^</span>Ek<span class="op">/</span>(V<span class="op">^</span>Ek<span class="op">+</span>E50<span class="op">^</span>Ek)
  E<span class="op">*</span>rateContact
}

<span class="kw">curve</span>(<span class="kw">rateInfect</span>(x, mu), <span class="dv">0</span>, <span class="dv">7</span>, <span class="dt">col=</span><span class="st">"red"</span>)</code></pre></div>
<p><img src="UserGuide_files/figure-html/unnamed-chunk-8-1.png" width="336"></p>
</div>
<div id="within-host-strain-mutation-rate" class="section level3">
<h3 class="hasAnchor">
<a href="#within-host-strain-mutation-rate" class="anchor"></a>Within-host strain mutation rate</h3>
<p>We define a function <code>rateMutate</code> which calculates the within-host per-locus mutation rate for a number K of infected hosts. The function recieves the GEVs matrix as an argument and 3 vectors of length K as follows:</p>
<ul>
<li>es: numeric : values of currently active host-specific effects;</li>
<li>envs: integer : host types for the K infected hosts;</li>
<li>genes: integer : currently infecting strains for the K hosts.</li>
</ul>
<p>In addition, we specify the mode of within-host evolution (in this case - selection)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># per locus mutation rates</span>
rateMutate &lt;-<span class="st"> </span><span class="cf">function</span>(GEValues, es, envs, genes) {
  z &lt;-<span class="st"> </span>GEValues[<span class="kw">cbind</span>(envs, genes)] <span class="op">+</span><span class="st"> </span>es
  V &lt;-<span class="st"> </span><span class="dv">10</span><span class="op">^</span>z
  Mmin &lt;-<span class="st"> </span><span class="fl">0.00</span>
  Mmax &lt;-<span class="st"> </span><span class="fl">0.2</span>
  M50 &lt;-<span class="st"> </span><span class="dv">10</span><span class="op">^</span><span class="dv">3</span>
  Mk &lt;-<span class="st"> </span><span class="fl">1.4</span>
  Mmin<span class="op">+</span>(Mmax<span class="op">-</span>Mmin)<span class="op">*</span>V<span class="op">^</span>Mk<span class="op">/</span>(V<span class="op">^</span>Mk<span class="op">+</span>M50<span class="op">^</span>Mk)
}

<span class="co"># are only beneficial (i.e. increasing the trait-value) mutations allowed</span>
selectWithinHost &lt;-<span class="st"> </span><span class="ot">TRUE</span></code></pre></div>
</div>
<div id="sampling-rate-for-an-infected-host" class="section level3">
<h3 class="hasAnchor">
<a href="#sampling-rate-for-an-infected-host" class="anchor"></a>Sampling rate for an infected host</h3>
<p>The event of sampling represents diagnosis and immediate treatment and lifelong immunity and un-infectiousness for the host. For this example, we define a constant sampling rate, independent of the phenotype.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rateSample &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">/</span>(<span class="dv">4</span><span class="op">*</span><span class="dv">12</span>)</code></pre></div>
</div>
</div>
<div id="define-the-time-step-and-the-duration-of-the-simulation" class="section level2">
<h2 class="hasAnchor">
<a href="#define-the-time-step-and-the-duration-of-the-simulation" class="anchor"></a>Define the time step and the duration of the simulation</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># all events are sampled with this time-step. This means that only one event can happen for an infected host within every next interval of 0.05 (arbitrary time units).</span>
timeStep &lt;-<span class="st"> </span><span class="fl">0.05</span>

<span class="co"># maximum time before starting graceful fadeout of the epidemic (stop the</span>
<span class="co"># transmission events and wait until no more infected hosts live in the population)</span>
maxTime &lt;-<span class="st"> </span><span class="dv">200</span>

<span class="co"># continue the epidemic outbreak until reaching maxNTips diagnosed hosts</span>
maxNTips &lt;-<span class="st"> </span><span class="dv">1000</span>

<span class="co"># time to continue the simulation of transmission after reaching maxNTips </span>
<span class="co"># (this was introduced in order to study post-outbreak dynamics, i.e. epidemic</span>
<span class="co"># waves after exhaustion of the susceptible pool)</span>
expandTimeAfterMaxNTips &lt;-<span class="st"> </span><span class="dv">0</span></code></pre></div>
</div>
</div>
<div id="running-the-simulation" class="section level1">
<h1 class="hasAnchor">
<a href="#running-the-simulation" class="anchor"></a>Running the simulation</h1>
<p>To run the simulation we use the function simulateEpidemic with the parameters specified as above:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">epidemic &lt;-<span class="st"> </span><span class="kw"><a href="../reference/simulateEpidemic.html">simulateEpidemic</a></span>(
  <span class="dt">Ninit=</span>N, <span class="dt">nu=</span>nu, <span class="dt">mu=</span>mu, <span class="dt">pe=</span>pe, <span class="dt">sde=</span>sde, <span class="dt">pg.init=</span>pg.init, <span class="dt">GEValues=</span>GEVs,
  <span class="dt">rateContact=</span><span class="dv">1</span><span class="op">/</span><span class="dv">6</span>, <span class="dt">rateInfect=</span>rateInfect, <span class="dt">rateDie=</span>rateDie,
  <span class="dt">rateSample=</span>rateSample,
  <span class="dt">rateMutate=</span>rateMutate,
  <span class="dt">numsAllelesAtSites=</span>numsAllelesAtSites, <span class="dt">eUniqForEachG=</span>eUniqForEachG,
  <span class="dt">selectWithinHost=</span><span class="ot">TRUE</span>,
  <span class="dt">timeStep=</span>timeStep, <span class="dt">maxTime=</span>maxTime, <span class="dt">maxNTips=</span>maxNTips,
  <span class="dt">expandTimeAfterMaxNTips=</span>expandTimeAfterMaxNTips,
  <span class="dt">process=</span><span class="st">"select/select"</span>)</code></pre></div>
</div>
<div id="analysing-the-epidemic" class="section level1">
<h1 class="hasAnchor">
<a href="#analysing-the-epidemic" class="anchor"></a>Analysing the epidemic</h1>
<p>Here we glimpse over some of the functions used in analyzing the simulated epidemic. A more elaborate analysis is provided in the vignette for the package <code>patherit</code>.</p>
<div id="extracting-the-transmission-tree-connecting-sampled-host" class="section level2">
<h2 class="hasAnchor">
<a href="#extracting-the-transmission-tree-connecting-sampled-host" class="anchor"></a>Extracting the transmission tree connecting sampled host</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tree &lt;-<span class="st"> </span><span class="kw"><a href="../reference/extractTree.html">extractTree</a></span>(epidemic)</code></pre></div>
<pre><code>## Generating tree: nTips= 1000 , number of uncollapsed edges= 3393</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(ape<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/ape/topics/ladderize">ladderize</a></span>(tree), <span class="dt">show.tip.label =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p><img src="UserGuide_files/figure-html/unnamed-chunk-14-1.png" width="672"></p>
</div>
<div id="extract-the-population-of-sampled-patients" class="section level2">
<h2 class="hasAnchor">
<a href="#extract-the-population-of-sampled-patients" class="anchor"></a>Extract the population of sampled patients</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># extract the population of all sampled individuals</span>
pop &lt;-<span class="st"> </span><span class="kw"><a href="../reference/extractPop.html">extractPop</a></span>(epidemic, <span class="dt">ids =</span> tree<span class="op">$</span>tip.label)

<span class="co"># calculate their phenotypic values at the moment of diagnosis</span>
pop[, z<span class="op">:</span><span class="er">=</span><span class="kw"><a href="../reference/calcValue.html">calcValue</a></span>(env, gene, e, <span class="dt">GEValues =</span> epidemic<span class="op">$</span>GEValues)]

<span class="kw">hist</span>(pop[, z])</code></pre></div>
<p><img src="UserGuide_files/figure-html/unnamed-chunk-15-1.png" width="672"></p>
</div>
<div id="extract-donor-recipient-couples" class="section level2">
<h2 class="hasAnchor">
<a href="#extract-donor-recipient-couples" class="anchor"></a>Extract donor-recipient couples</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">drc1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/extractDRCouples.html">extractDRCouples</a></span>(epidemic)</code></pre></div>
</div>
<div id="saving-the-epidemic-for-future-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#saving-the-epidemic-for-future-analysis" class="anchor"></a>Saving the epidemic for future analysis</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">save</span>(epidemic, <span class="dt">file=</span><span class="st">"UserGuideCache.RData"</span>)</code></pre></div>
</div>
</div>
<div id="packages-used" class="section level1">
<h1 class="hasAnchor">
<a href="#packages-used" class="anchor"></a>Packages used</h1>
<p>Apart from base R functionality, the toyepidemic package uses a number of 3rd party R-packages:</p>
<ul>
<li>For fast execution: Rcpp v0.12.14 <span class="citation">(Eddelbuettel et al. 2017)</span>;</li>
<li>For tree processing: ape v5.0 <span class="citation">(Paradis et al. 2016)</span>;</li>
<li>For reporting: data.table v1.10.4.3 <span class="citation">(Dowle and Srinivasan 2016)</span>;</li>
<li>For testing: testthat v1.0.2 <span class="citation">(Wickham 2016)</span>.</li>
</ul>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-R-data-table">
<p>Dowle, Matt, and Arun Srinivasan. 2016. <em>Data.table: Extension of ‘Data.frame‘</em>. <a href="https://CRAN.R-project.org/package=data.table" class="uri">https://CRAN.R-project.org/package=data.table</a>.</p>
</div>
<div id="ref-R-Rcpp">
<p>Eddelbuettel, Dirk, Romain Francois, JJ Allaire, Kevin Ushey, Qiang Kou, Nathan Russell, Douglas Bates, and John Chambers. 2017. <em>Rcpp: Seamless R and C++ Integration</em>. <a href="https://CRAN.R-project.org/package=Rcpp" class="uri">https://CRAN.R-project.org/package=Rcpp</a>.</p>
</div>
<div id="ref-R-ape">
<p>Paradis, Emmanuel, Simon Blomberg, Ben Bolker, Julien Claude, Hoa Sien Cuong, Richard Desper, Gilles Didier, et al. 2016. <em>Ape: Analyses of Phylogenetics and Evolution</em>. <a href="https://CRAN.R-project.org/package=ape" class="uri">https://CRAN.R-project.org/package=ape</a>.</p>
</div>
<div id="ref-R-testthat">
<p>Wickham, Hadley. 2016. <em>Testthat: Unit Testing for R</em>. <a href="https://CRAN.R-project.org/package=testthat" class="uri">https://CRAN.R-project.org/package=testthat</a>.</p>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#designing-the-epidemic">Designing the epidemic</a><ul class="nav nav-pills nav-stacked">
<li><a href="#define-the-pathogen-genotypes">Define the pathogen genotypes</a></li>
      <li><a href="#define-the-number-and-the-frequencies-of-host-types">Define the number and the frequencies of host-types</a></li>
      <li><a href="#define-the-ge-values">Define the GE-values</a></li>
      <li><a href="#define-the-host-specific-effects-distribution">Define the host-specific effects distribution</a></li>
      <li><a href="#defining-the-population-parameters">Defining the population parameters</a></li>
      <li><a href="#define-the-time-step-and-the-duration-of-the-simulation">Define the time step and the duration of the simulation</a></li>
      </ul>
</li>
      <li><a href="#running-the-simulation">Running the simulation</a></li>
      <li>
<a href="#analysing-the-epidemic">Analysing the epidemic</a><ul class="nav nav-pills nav-stacked">
<li><a href="#extracting-the-transmission-tree-connecting-sampled-host">Extracting the transmission tree connecting sampled host</a></li>
      <li><a href="#extract-the-population-of-sampled-patients">Extract the population of sampled patients</a></li>
      <li><a href="#extract-donor-recipient-couples">Extract donor-recipient couples</a></li>
      <li><a href="#saving-the-epidemic-for-future-analysis">Saving the epidemic for future analysis</a></li>
      </ul>
</li>
      <li><a href="#packages-used">Packages used</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Venelin Mitov.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
